scala/providers.bzl

```diff
@@ -57,3 +57,10 @@ declare_scalac_provider = rule(
         "default_macro_classpath": attr.label_list(allow_files = True),
     },
 )
+
+ScalaRulePhase = provider(
+    doc = "A Scala compiler plugin",
+    fields = {
+        "phases": "the phases to add",
```

Comment:
More documentations and tests for sure.

Do you think there is a way to improve the double level of customization? Would you prefer we make it simpler?

Reply:
Re double level see comment in summary of review.
Concretely we need to document the vector `(relation, peer_name, name, function)` here since it's the API.
Also, should the field be called `customizable_phases` or `custom_phases`?

Reply:
I prefer `custom_phases` to keep it shorter. I will update the PR.